addons/udes_stock_refactoring/models/stock_picking.py

```diff
@@ -0,0 +1,176 @@
+
+_logger = logging.getLogger(__name__)
+
+
+class StockPicking(models.Model):
+    _inherit = "stock.picking"
+
+    u_mark = fields.Boolean(
+        "Marked",
+        default=True,
+        index=True,
+        help="Pickings that are unused after refactoring are unmarked and deleted",
+    )
+
+    def find_empty_pickings_to_unlink(self):
+        """
+        Finds empty pickings to unlink in self, when it is set, otherwise
+        searches for any empty picking.
+
+        Filter out of the generic search when the picking type does not
+        have auto unlink empty enabled.
+        """
+        PickingType = self.env["stock.picking.type"]
+
+        domain = [
+            ("u_mark", "=", False),
+            ("is_locked", "=", True),
+        ]
+        if self:
+            domain.append(("id", "in", self.ids))
+        else:
+            pts = PickingType.search([("u_auto_unlink_empty", "=", False)])
+            if pts:
+                domain.append(("picking_type_id", "not in", pts.ids))
+        while True:
+            records = self.search(domain, limit=1000)
+            if not records:
+                break
+            yield records
+
+    @api.model
+    def _new_picking_for_group(self, group_key, moves, **kwargs):
+        """
+        Find existing picking for the supplied group, if none found create a new one.
+        Assign the moves to the picking and return it.
+        """
+        ProcurementGroup = self.env["procurement.group"]
+
+        picking_type = moves.mapped("picking_type_id")
+        picking_type.ensure_one()
+        src_loc = moves.mapped("location_id")
+        dest_loc = moves.mapped("location_dest_id")
+
+        group = ProcurementGroup.get_or_create(group_key, create=True)
+
+        # Look for an existing picking with the right group
+        picking = self.search(
+            [
+                ("picking_type_id", "=", picking_type.id),
+                ("location_id", "=", src_loc.id),
+                ("location_dest_id", "=", dest_loc.id),
+                ("group_id", "=", group.id),
+                # NB: only workable pickings
+                ("state", "in", ["assigned", "confirmed", "waiting"]),
+            ]
+        )
+
+        if not picking:
+            # Otherwise reuse the existing picking if all the moves
+            # already belong to it and it contains no other moves
+            # The picking_type_id, location_id and location_dest_id
+            # will match already
+            current_picking = moves.mapped("picking_id")
+            if len(current_picking) == 1 and current_picking.mapped("move_lines") == moves:
+                values = {"group_id": group.id}
+                values.update(kwargs)
+
+                current_picking.write(values)
+                picking = current_picking
+
+        if not picking or len(picking) > 1:
+            # There was no suitable picking to reuse.
+            # Create a new picking.
+            values = {
```

Comment:
should we get this default set of values via a method?

Reply:
Have added a `_get_default_new_picking_for_group_values` method for this. This method is now also used in conjunction with the `create_picking` helper method from `udes_stock`.

I've also split out the functionality in this method for removing misleading values from the picking into its own `_remove_misleading_values` method.